### PR TITLE
SDK Connection Strings (Net Core support)

### PR DESCRIPTION
**Pull request by TimothyMothra**
https://github.com/microsoft/ApplicationInsights-dotnet/issues/1221

**TODO**
- Need to discuss Javascript Snippit with SDK team.

**Changes**
- Extensions and ServiceOptions

### CODE REVIEW

File: test/Microsoft.ApplicationInsights.AspNetCore.Tests/Extensions/ApplicationInsightsExtensionsTests.cs [hunk 1]
```diff
@@ -201,15 +215,38 @@ public static void RegistersTelemetryConfigurationFactoryMethodThatReadsInstrume
...
+            public static void AddApplicationInsightsTelemetry_ReadsConnectionString_FromEnvironment()
+            {
+                var services = ApplicationInsightsExtensionsTests.GetServiceCollectionWithContextAccessor();
+                Environment.SetEnvironmentVariable(ConnectionStringEnvironmentVariable, TestConnectionString);
+                try
+                {
+                    services.AddApplicationInsightsTelemetry();
+                    IServiceProvider serviceProvider = services.BuildServiceProvider();
+                    var telemetryConfiguration = serviceProvider.GetTelemetryConfiguration();
+                    Assert.Equal(TestConnectionString, telemetryConfiguration.ConnectionString);
```
Comment (cijothomas) [inline test/Microsoft.ApplicationInsights.AspNetCore.Tests/Extensions/ApplicationInsightsExtensionsTests.cs]: validate that config.ikey, endpoint etc are also set from con string.

File: test/Microsoft.ApplicationInsights.AspNetCore.Tests/Extensions/ApplicationInsightsExtensionsTests.cs [hunk 2]
```diff
@@ -201,15 +215,38 @@ public static void RegistersTelemetryConfigurationFactoryMethodThatReadsInstrume
...
             /// <summary>
             /// Validates that while using services.AddApplicationInsightsTelemetry(); ikey is read from
             /// Environment
             /// </summary>
             [Fact]
             public static void AddApplicationInsightsTelemetryReadsInstrumentationKeyFromEnvironment()
             {
-                var services = ApplicationInsightsExtensionsTests.GetServiceCollectionWithContextAccessor();                
-                Environment.SetEnvironmentVariable("APPINSIGHTS_INSTRUMENTATIONKEY", TestInstrumentationKey);                
+                var services = ApplicationInsightsExtensionsTests.GetServiceCollectionWithContextAccessor();
```
Comment (cijothomas) [inline test/Microsoft.ApplicationInsights.AspNetCore.Tests/Extensions/ApplicationInsightsExtensionsTests.cs]: lets add one test with both ikey,endpoint in appsettings.json AND connectionstring also on appsettings.json, Validate that connectionstring Wins.

File: src/Shared/Implementation/TelemetryConfigurationOptionsSetup.cs [hunk 3]
```diff
@@ -65,6 +65,11 @@ public void Configure(TelemetryConfiguration configuration)
                     configuration.InstrumentationKey = this.applicationInsightsServiceOptions.InstrumentationKey;
                 }
 
+                if (this.applicationInsightsServiceOptions.ConnectionString != null)
```
Comment (cijothomas) [inline src/Shared/Implementation/TelemetryConfigurationOptionsSetup.cs]: we read endpoint address further down this class. That should be moved before ConString. Else it'll override connectionstring's endpoint address.

File: src/Shared/Implementation/TelemetryConfigurationOptionsSetup.cs [hunk 4]
```diff
@@ -117,6 +117,12 @@ public void Configure(TelemetryConfiguration configuration)
                     configuration.TelemetryChannel.EndpointAddress = this.applicationInsightsServiceOptions.EndpointAddress;
                 }
 
+                // Need to set connection string before calling Initialize() on the Modules and Processors.
+                if (this.applicationInsightsServiceOptions.ConnectionString != null)
+                {
+                    configuration.ConnectionString = this.applicationInsightsServiceOptions.ConnectionString;
```
Comment (cijothomas) [inline src/Shared/Implementation/TelemetryConfigurationOptionsSetup.cs]: will this overwrite the channel.endpoint as well?

File: src/Shared/Extensions/ApplicationInsightsServiceOptions.cs [hunk 5]
```diff
@@ -164,6 +169,7 @@ internal void CopyPropertiesTo(ApplicationInsightsServiceOptions target)
                 target.InstrumentationKey = this.InstrumentationKey;
             }
 
+            target.ConnectionString = this.ConnectionString;
```
Comment (cijothomas) [inline src/Shared/Extensions/ApplicationInsightsServiceOptions.cs]: do you need to check constring for null or empty?

**Comment by TimothyMothra** [comment]
I didn't see your comments until after I clicked merge. I'm reviewing now.

File: src/Shared/Extensions/ApplicationInsightsServiceOptions.cs (hunk 5 above)
Comment (TimothyMothra) [reply src/Shared/Extensions/ApplicationInsightsServiceOptions.cs]: this isn't necessary. it won't affect the behavior to copy the properties as is from one ServiceOptions to the other

File: src/Shared/Implementation/TelemetryConfigurationOptionsSetup.cs (hunk 4 above)
Comment (TimothyMothra) [reply src/Shared/Implementation/TelemetryConfigurationOptionsSetup.cs]: Yes, which is why I put it last.

**Pull request merged**